libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -98,13 +98,20 @@ void IRGenerationContext::registerImmutableVariable(VariableDeclaration const& _
 	*m_reservedMemory += _variable.annotation().type->memoryHeadSize();
 }
 
-size_t IRGenerationContext::immutableMemoryOffset(VariableDeclaration const& _variable) const
+size_t IRGenerationContext::immutableMemoryOffset(VariableDeclaration const& _variable, bool _relative) const
 {
 	solAssert(
 		m_immutableVariables.count(&_variable),
 		"Unknown immutable variable: " + _variable.name()
 	);
-	return m_immutableVariables.at(&_variable);
+	auto const absoluteOffset = m_immutableVariables.at(&_variable);
+	return !_relative ? absoluteOffset : (absoluteOffset - CompilerUtils::generalPurposeMemoryStart);
+}
+
+size_t IRGenerationContext::immutablesMemorySize() const
```

Comment:
Technically, this could be used for any pre-allocated memory the compiler may need, not just the bits reserved for immutables, so the name is too specific. Let's stick to the convention we're already using for it:

```suggestion
size_t IRGenerationContext::reservedMemorySize() const
```

Reply:
Also, if anything other than immutables was to be stored in it, it would be bad, because the current mechanism would put it in auxdata along with immutables. I guess it's fine for now, since it looks like we don't really do that, but we definitely need a sanity check that will fail if someone tries to do it. You should make `IRGenerationContext::reservedMemory()` assert that the size of reserved memory matches the size of all immutables (with a comment explaining why this assumption is necessary).

The size of immutables would be simply `m_immutableVariables.size() * 32`, but that's under the assumption that they are all value types. Since we're going to change that eventually, may be better to actually iterate over the variables and get their actual size. Or at least leave a TODO to do it when we introduce such immutables.

Reply:
Done